grid2op/Opponent/OpponentSpace.py

```diff
@@ -103,18 +129,51 @@ def attack(self, observation, agent_action, env_action):
 
         Returns
         -------
-        res: :class:`grid2op.Action.Action`
-            The attack the opponent wants to perform (or "do nothing" if the attack was too costly)
+        res: :class:`grid2op.Action.Action` : The attack the opponent wants to perform
+                                              (or "do nothing" if the attack was too costly)
+              or class:`NoneType` : Returns None if no action is taken
 
         """
+
+        if observation is None:
+            # this is the first time step, which is not a "real" one
+            # just here to load the data properly, so opponent do not attack there
+            return None, 0
+
+        # Update variables
         self.budget += self.budget_per_timestep
-        attack = self.opponent.attack(observation, agent_action, env_action, self.budget, self.previous_fails)
-        tmp = self.compute_budget(attack)
-        if tmp <= self.budget:
-            self.previous_fails = False
-            self.budget -= tmp
-            res = attack
+        self.current_attack_duration = max(0, self.current_attack_duration - 1)
+        self.current_attack_cooldown = max(0, self.current_attack_cooldown - 1)
+
+        # If currently attacking
+        if self.current_attack_duration > 0:
+            attack = self.last_attack
+            self.opponent.tell_attack_continues(observation, agent_action, env_action, self.budget)
+
+        # If the opponent has already attacked today
+        elif self.current_attack_cooldown > self.attack_cooldown:
+            attack = None
+
+        # If the opponent can attack  
         else:
-            self.previous_fails = True
-            res = self._do_nothing
-        return res
+            self.previous_fails = False
+            attack = self.opponent.attack(observation, agent_action, env_action, self.budget,
+                                          self.previous_fails)
+            # If the cost is too high
+            if self.attack_duration * self.compute_budget(attack) > self.budget:
+                attack = None
+                self.previous_fails = True
```

Comment:
Shouldn't it call self.has_failed() instead?

Reply:
And maybe have a `self.has_succeeded()` call in the other cases, instead of `self.previous_fails = False` before the `if` statement

Reply:
has_failed does not exists in the opponent API.

Reply:
That being said, thanks for pointing this out. When you coded this part, you forgot to call `tell_attack_continues` when... the attack continues. So the opponent is NOT updated in the case :-/